clippy_lints/src/misc_early.rs

```diff
@@ -391,92 +390,102 @@ impl EarlyLintPass for MiscEarlyLints {
 
 impl MiscEarlyLints {
     fn check_lit(self, cx: &EarlyContext<'_>, lit: &Lit) {
-        if_chain! {
-            if let LitKind::Int(value, ..) = lit.node;
-            if let Some(src) = snippet_opt(cx, lit.span);
-            if let Some(firstch) = src.chars().next();
-            if char::to_digit(firstch, 10).is_some();
-            then {
-                let mut prev = '\0';
-                for (idx, ch) in src.chars().enumerate() {
-                    if ch == 'i' || ch == 'u' {
-                        if prev != '_' {
-                            span_lint_and_sugg(
-                                cx,
-                                UNSEPARATED_LITERAL_SUFFIX,
-                                lit.span,
-                                "integer type suffix should be separated by an underscore",
-                                "add an underscore",
-                                format!("{}_{}", &src[0..idx], &src[idx..]),
-                                Applicability::MachineApplicable,
-                            );
-                        }
-                        break;
-                    }
-                    prev = ch;
+        // The `line!()` macro is compiler built-in and a special case for these lints.
+        let lit_snip = match snippet_opt(cx, lit.span) {
+            Some(snip) => {
+                if snip.contains('!') {
```

Comment:
That's an interesting hack 😄 Does this also work if you check for `line!` explicitly?
```suggestion
                if snip.contains('line!') {
```

Also:
```rust
let lit_snip = match snippet_opt(cx, lit.span) {
    Some(snip) if !snip.contains('!') => snip,
    _ => return,
}
```

Reply:
The code above will require Bind-By-Move feature [issue][1].
Also we the check for `line` is not enough, because `snippet_out`
could return `assert!(_)` which we would also check too.

[1]: https://github.com/rust-lang/rust/issues/15287

Reply:
```rust
let lit_snip = match &snippet_opt(cx, lit.span) {
    Some(snip) if !snip.contains('!') => *snip,
    _ => return,
}
```

Matching a reference of `snippet_opt` should work though?

Reply:
This returns error: "temporary value dropped while borrowed"